src/librustc_mir/util/pretty.rs

```diff
@@ -597,7 +597,8 @@ fn write_mir_sig(
     trace!("write_mir_sig: {:?}", src.instance);
     let descr = tcx.describe_def(src.def_id());
     let is_function = match descr {
-        Some(Def::Fn(_)) | Some(Def::Method(_)) | Some(Def::StructCtor(..)) => true,
+        Some(Def::Fn(_)) | Some(Def::Method(_)) |
+        Some(Def::StructCtor(..)) | Some(Def::Variant(..)) => true,
```

Comment:
You need `Def::VariantCtor` here rather than `Def::Variant`, `Def::Variant` is the "variant type" from type namespace.
Also, only `Def::VariantCtor(_, CtorKind::Fn)` is a function, same applies to `Def::StructCtor`.

Reply:
Ah, you're right. Good catch

Reply:
@petrochenkov Actually I spoke too soon. The ICE is still hit when `Def::VariantCtor` is used.

I think we end up with the `Def::Variant` here because in [`tcx.describe_def`](https://github.com/rust-lang/rust/blob/master/src/librustc/hir/map/mod.rs#L313) there is no arm that will return a `VariantCtor`. I think we use the same `DefId` for the ctor and the variant.

Reply:
>in tcx.describe_def there is no arm that will return a VariantCtor. I think we use the same DefId for the ctor and the variant.

Sigh, ok.
This is not a good situation, but variants sharing def-ids with their is a long standing issue.